src/frequenz/sdk/actor/power_distributing/_distribution_algorithm/_distribution_algorithm.py

```diff
@@ -304,113 +421,186 @@ def _distribute_power(  # pylint: disable=too-many-arguments
+        deficits: dict[InverterSet, float] = {}
+
+        for ratio_data in battery_availability_ratio:
             # ratio = 0, means all remaining batteries reach max SoC lvl or have no
             # capacity
             if is_close_to_zero(ratio):
-                distribution[inverter.component_id] = 0.0
+                distribution[InverterSet(ratio_data.inverter_ids)] = _Power(
+                    upper_bound=0.0,
+                    power=0.0,
+                )
                 continue
 
             power_to_distribute = power_w - reserved_power
-            calculated_power = power_to_distribute * battery_ratio / ratio
-            reserved_power += max(calculated_power, excl_bound)
-            used_ratio += battery_ratio
+            calculated_power = power_to_distribute * ratio_data.ratio / ratio
+            reserved_power += max(calculated_power, ratio_data.min_power)
+            used_ratio += ratio_data.ratio
             ratio = sum_ratio - used_ratio
-            # If the power allocated for that inverter is out of bound,
+
+            # If the power allocated for that inverter set is out of bound,
             # then we need to distribute more power over all remaining batteries.
-            incl_bound = incl_bounds[inverter.component_id]
+            incl_bound = min(
+                sum(
+                    incl_bounds[inverter_id] for inverter_id in ratio_data.inverter_ids
+                ),
+                incl_bounds[ratio_data.battery_id],
+            )
             if calculated_power > incl_bound:
-                excess_reserved[inverter.component_id] = incl_bound - excl_bound
-                # # Distribute between remaining batteries
-            elif calculated_power < excl_bound:
-                deficits[inverter.component_id] = calculated_power - excl_bound
+                excess_reserved[frozenset(ratio_data.inverter_ids)] = (
+                    incl_bound - ratio_data.min_power
+                )
+            # # Distribute between remaining batteries
+            elif calculated_power < ratio_data.min_power:
+                deficits[frozenset(ratio_data.inverter_ids)] = (
+                    calculated_power - ratio_data.min_power
+                )
             else:
-                excess_reserved[inverter.component_id] = calculated_power - excl_bound
+                excess_reserved[frozenset(ratio_data.inverter_ids)] = (
+                    calculated_power - ratio_data.min_power
+                )
+
+            distributed_power += ratio_data.min_power
+            distribution[InverterSet(ratio_data.inverter_ids)] = _Power(
+                upper_bound=incl_bound,
+                power=ratio_data.min_power,
+            )
 
-            distributed_power += excl_bound
-            distribution[inverter.component_id] = excl_bound
+        make_largest = make_dataclass(
+            "make_largest", [("inverter_ids", frozenset), ("power", float)]
+        )
```

Comment:
Does mypy understand this syntax?  Why not just the normal one?

Also I think we shouldn't do this as part of the hot-path, better to move it outside the function.  Because it is still a runtime step in Python.

```suggestion
@dataclass
class _Allocation:
    inverter_ids: InverterSet
    power: float